spec/unit/network/http/pool_spec.rb

```diff
@@ -166,6 +170,16 @@ def create_connection(site)
 
         pool.with_connection(site, verify) {|c| }
       end
+
+      it "doesn't add a closed  connection back to the pool" do
+        conn = create_connection(site)
+
+        pool = create_pool_with_connections(site, conn)
+
+        expect(pool).not_to receive(:release).with(site, conn)
+
+        pool.with_connection(site, verify) {|c| }
```

Comment:
I think you'll want to call `{|c| c.close }`, and then verify that the pool is empty after this call (as opposed to release never being called).

Reply:
I'm a little confused by this.

In the code, we yield to the block, and then after, call ensure to close the connection in the `ensure`.

I agree checking that `release` was never called isn't a good test. But I'm having trouble figuring out exactly how to verify the pool is empty.

I think we still want `pool.with_connection(site, verify) {|c| }`.

Before that call, pool is
```
=> #<Puppet::Network::HTTP::Pool:0x00007fe91517dad0
 @factory=#<Puppet::Network::HTTP::Factory:0x00007fe91517da80>,
 @keepalive_timeout=15,
 @pool=
  {#<Puppet::Network::HTTP::Site:0x00007fe91517ea20 @host="rubygems.org", @port=443, @scheme="https">=>
    [#<Puppet::Network::HTTP::Session:0x00007fe91517da08
      @connection=#<Double "https://rubygems.org:443">,
      @expiration_time=2020-02-07 10:29:43 -0800>]}>
```

And after it's
```
=> #<Puppet::Network::HTTP::Pool:0x00007fe91517dad0
 @factory=#<Puppet::Network::HTTP::Factory:0x00007fe91517da80>,
 @keepalive_timeout=15,
 @pool={#<Puppet::Network::HTTP::Site:0x00007fe91517ea20 @host="rubygems.org", @port=443, @scheme="https">=>[]}>
```

`pool.pool` gets me closer
```
=> {#<Puppet::Network::HTTP::Site:0x00007fde21c667c0 @host="rubygems.org", @port=443, @scheme="https">=>[]}
```

but that feels wrong.

Also
```
> pool.pool.empty?
=> false
```

Reply:
> In the code, we yield to the block, and then after, call ensure to close the connection in the ensure.

As far as I can tell, the call to `http.finish` happens during the yield before the ensure block. So something like

```
with_connection
-> conn = borrow
-> begin
  -> yield conn
    -> execute_request(conn)
    -> conn.finish
-> ensure
  -> if reuse && !conn.active?
```

> pool.pool gets me closer

```=> {#<Puppet::Network::HTTP::Site:0x00007fde21c667c0 @host="rubygems.org", @port=443, @scheme="https">=>[]}
```

Yep, sorry we really care that there aren't any active connections for that site. I was conflating that with the pool being empty.

So that looks like a minor bug in the pool, as there is a site that is referring to an empty array of connections. If `close_connection` is called and there are no more active connections for that site then we should delete the site from the pool. Otherwise it's just taking up memory and will never be gc'ed, or until the agent exits, which it may never for daemonized agents. I say minor because the next time the agent runs, it will reuse the same site entry. Could you file an issue on that?

Reply:
Are we looking at the same code? Or is there a `finish` call that happens in the block that's passed to `with_connection`? I feel like I'm missing something.
In `with_connection`, this is in the ensure block: 
```
      if reuse
        release(site, http)
      else
        close_connection(site, http)
      end
```
and this is `close_connection` that it calls
```
  def close_connection(site, http)
    Puppet.debug("Closing connection for #{site}")
    http.finish
  rescue => detail
    Puppet.log_exception(detail, _("Failed to close connection for %{site}: %{detail}") % { site: site, detail: detail })
  end
```

The only thing that calls `with_connection` and also calls finish in the block is in `do_request`, in the case where we get a 429 or 503 error code, which I added as a part of this PR.